### PR TITLE
Implemented the RegisterOnTermination feature.

### LINKED ISSUE #1519: Implement RegisterOnTermination
In the [scala](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/ActorSystem.scala#L376) version it's possible to register callbacks which will be executed after all actors have been terminated. 
This is also referenced in the [Shutdown](https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka/Actor/ActorSystem.cs#L144) comment but not implemented. 

See also ActorSystemSpec #60

### PR DESCRIPTION
**Pull request by Silv3rcircl3**
closes #1519 
It also implements some of the tests for #60.

### CODE REVIEW

File: src/core/Akka/Util/AtomicReference.cs [hunk 1]
```diff
@@ -81,6 +81,18 @@ public bool CompareAndSet(T expected, T newValue)
             return false;
         }
 
+        /// <summary>
+        /// Returns the current value and overrides it with the new one.
+        /// </summary>
+        /// <param name="newValue">The new value</param>
+        /// <returns>The current value</returns>
+        public T GetAndSet(T newValue)
```
Comment (Horusiath) [inline src/core/Akka/Util/AtomicReference.cs]: This implementation doesn't seems to be atomic.

Comment (Silv3rcircl3) [inline src/core/Akka/Util/AtomicReference.cs]: Yeah I struggled with this one, thanks for the hint, will update the PR.

Comment (Silv3rcircl3) [inline src/core/Akka/Util/AtomicReference.cs]: ```Interlocked.Exchange``` unfortunately only allows reference types  in the generic version.
Would this also work?
```
Monitor.Enter(_lock);
var current = Value;
Value = newValue;
Monitor.Exit(_lock);
return current;
```

**Comment by Aaronontheweb** [comment]
Looks great! How does this integrate with the termination hooks for Akka.Remote? Or does this live as a separate system?

https://github.com/akkadotnet/akka.net/blob/efdacf8bf9d179956d5af16269b837fb2dd70d57/src/core/Akka.Remote/RemoteActorRefProvider.cs#L440

If the `RemotingTerminator` needs to be modified, we can do that as a separate issue.

**Comment by Silv3rcircl3** [comment]
Sorry, but through my low understanding about Akka.Remote I can't give an appropriate answer.

**Comment by Aaronontheweb** [comment]
@Silv3rcircl3 ok, no worries Marc - I'll add that as something for me to investigate

**Pull request merged**